File: apps/server/internal/domain/sensor_value_interpreter_test.go
package domain

import (
	"encoding/json"
	"testing"
)

func TestInterpretSensorSampleValueUsesGasStrategy(t *testing.T) {
	readings := InterpretSensorSampleValue(
		SensorDescriptor{
			SensorID:   "telemetry.gas.channel_1",
			Label:      "CO",
			SensorType: "gas",
			Unit:       "ppm",
		},
		SensorSample{
			Values: json.RawMessage(`{"alarm":"normal","concentration":35,"unit":"ppm","valid":true}`),
		},
	)

	if len(readings) != 1 {
		t.Fatalf("expected one gas reading, got %#v", readings)
	}
	if readings[0].Label != "CO" || readings[0].FieldPath != "concentration" || readings[0].Value != float64(35) {
		t.Fatalf("unexpected gas reading: %#v", readings[0])
	}
	if readings[0].Order != 10.01 || readings[0].Unit != "ppm" {
		t.Fatalf("unexpected gas presentation metadata: %#v", readings[0])
	}
}

func TestInterpretSensorSampleValueFlattensDefaultObject(t *testing.T) {
	readings := InterpretSensorSampleValue(
		SensorDescriptor{
			SensorID:   "telemetry.battery_1",
			Label:      "Battery",
			SensorType: "battery",
			Unit:       "percent",
		},
		SensorSample{
			Values: json.RawMessage(`{"batteryPercent":91,"positionAvailable":true}`),
		},
	)

	if len(readings) != 1 {
		t.Fatalf("expected one visible reading, got %#v", readings)
	}
	if readings[0].Label != "배터리" || readings[0].FieldPath != "batteryPercent" || readings[0].Value != float64(91) {
		t.Fatalf("unexpected battery reading: %#v", readings[0])
	}
	if readings[0].Order != 30 || readings[0].Unit != "%" {
		t.Fatalf("unexpected battery presentation metadata: %#v", readings[0])
	}
}

File: apps/web/src/domains/live/sensorValueInterpreter.test.js
import { describe, expect, it } from "vitest";
import { interpretSensorSampleValue } from "./sensorValueInterpreter.js";

describe("sensorValueInterpreter", () => {
  it("uses gas strategy for concentration readings", () => {
    const readings = interpretSensorSampleValue(
      {
        label: "CO",
        sensorId: "telemetry.gas.channel_1",
        sensorType: "gas",
        unit: "ppm"
      },
      {
        values: {
          alarm: "normal",
          concentration: 35,
          unit: "ppm",
          valid: true
        }
      }
    );

    expect(readings).toHaveLength(1);
    expect(readings[0]).toMatchObject({
      fieldPath: "concentration",
      label: "CO",
      unit: "ppm",
      value: 35
    });
  });

  it("flattens default object values and hides metadata fields", () => {
    const readings = interpretSensorSampleValue(
      {
        label: "Battery",
        sensorId: "telemetry.battery_1",
        sensorType: "battery",
        unit: "percent"
      },
      {
        values: {
          batteryPercent: 91,
          positionAvailable: true
        }
      }
    );

    expect(readings).toHaveLength(1);
    expect(readings[0]).toMatchObject({
      fieldPath: "batteryPercent",
      label: "배터리",
      unit: "%",
      value: 91
    });
  });
});
